### ISSUE TITLE
Display inherited values (via `crossref`)

### ISSUE THREAD

**Issue by apc** [bdarcus/citar#545]
I like to keep my bib entries as brief as I can and rely on crossref as much as possible. BibTeX and Biblatex help a lot here, and you do not need e.g. to enter a value for `booktitle` in an `InCollection` entry that has a non-empty `crossref` field. More relevant to the current request, Biblatex allows you not to specify a value for `author` in an `InBook` entry that has a non-empty `crossref` field. Unsurprisingly, perhaps, `citar` does not seem to have access to the information that Biblatex and BibTeX make available via crossref. 

Here's an example pair of entries: 

```tex
@Book{crossref,
  author       = {Smithers, Waylon},
  title        = {A Book},
  date         = 2020
}

@InBook{crossref2,
  title        = {A chapter},
  date         = 2020,
  crossref     = {crossref}
}
```

If I use `citar` to search for 'Smithers', the entry with key 'crossref2' will not show up. And if I search by key, e.g. 'crossref', the entry for 'crossref2' appears with no author: 

<img width="1267" alt="image" src="https://user-images.githubusercontent.com/350504/151865047-d670e1c5-06f7-45a6-8415-f302e55abed7.png">

Would it be possible for `citar` to parse a bib file and populate the candidates using crossref information? This may well be beyond the scope of this package, of course, but I thought I'd ask. 

I'm tweaking with a function that automatically populates fields that would otherwise be inherited via crossref. But I do not like this solution, since it would increase the chances of mistakes---e.g. updating the author name in an entry wouldn't automatically update the author name in any entry that is 'linked' (via crossref) to the first.

**Comment by bdarcus** [comment]
I defer all parsing to parsebib. Perhaps there are options for that there you could check on?

**Comment by apc** [comment]
Interesting. I think Ebib is pretty good at handling crossref information, so parsebib should be able to handle that...

**Comment by bdarcus** [comment]
> Interesting. I think Ebib is pretty good at handling crossref information, so parsebib should be able to handle that...

Likely, but not necessarily. In #535, a function that should likely be in parsebib is implemented in ebib.

But I'd guess you're right.

**Comment by apc** [comment]
I'm confused now. Looking at the documentation it seems as if `parsebib-parse`, which is I believe the function being called by `citar`, should handle crossref entries as I expected. If I run `parsebib-parse` on a file containing only the following two entries: 

```tex
@Book{key1,
  author       = {Smithers, Waylon},
  title        = {A Book},
  date         = 2020
}

@InBook{key2,
  title        = {A chapter},
  crossref     = {key1}
}
```

I get the following: 

```el
#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold
 0.8125 data ("key1" (("date" . "2020") ("title" . "A Book") ("author"
 . "Smithers, Waylon") ("=type=" . "Book") ("=key=" . "key1")) "key2"
 (("crossref" . "key1") ("title" . "A chapter") ("=type=" . "InBook")
 ("=key=" . "key2") ("date" . "2020") ("booktitle" . "A Book")
 ("author" . "Smithers, Waylon"))))
```

So it seems as if parsebib is populating the list with the information inherited via crossref. 

If I call `citar--format-candidates` with that same file, I get the following: 

```el
((#("                  A chapter                                            key2               InBook               key2" 0 65 (face citar-highlight) 66 106 (face citar) 107 115 (invisible t)) "key2" ("title" . "A chapter") ("=type=" . "InBook") ("=key=" . "key2")) (#("Smithers          A Book                                     2020      key1               Book                 key1" 0 65 (face citar-highlight) 66 106 (face citar) 107 115 (invisible t)) "key1" ("date" . "2020") ("title" . "A Book") ("author" . "Smithers, Waylon") ("=type=" . "Book") ("=key=" . "key1")))
```

I do not understand well enough how any of this is working under the hood to see why this is happening, since from what I can tell `citar--format-candidates` is getting the information to build the list of candidates from `parsebib-parse`. Thoughts?

**Comment by bdarcus** [comment]
I think you want to look into the inheritance arg in `parsebib-parse-bib-buffer`, which seems to be set in the display arg in `parsebib-parse`?

**Comment by aikrahguzar** [comment]
We parse an explicit list of fields to `parsebib-parse` and this might be because `"crossref"` is not among them. Try 
```elisp
(add-to-list 'citar-additional-fields "crossref")
(citar-refresh)
```

**Comment by apc** [comment]
> I think you want to look into the inheritance arg in `parsebib-parse-bib-buffer`, which seems to be set in the display arg in `parsebib-parse`?

@bdarcus: The display arg is by default set to `t`, from what I can tell:

```
Signature
(parsebib-parse FILES &key ENTRIES STRINGS (DISPLAY t) FIELDS)

Documentation
Parse one or more bibliography files.

FILES is the list of files to parse.  All bibliographic entries
in FILES are collected and returned in a single hash table.
FILES can be a list of .bib or .json files, or a combination
of these.  FILES can also be a string, which should be the path
to a single bibliography file.

ENTRIES, if provided, should be a hash table with test function
equal, it is used to store the entries.  Any existing entries
with identical keys are overwritten.  If provided, ENTRIES is
also the return value.  If ENTRIES is nil, a new hash table is
created and returned.

STRINGS, similarly a hash table with test function equal, is
used to store the @String definitions.  Although STRINGS is not
returned, it is modified in place and can therefore be used to
collect the @String definitions in the files being parsed.

If DISPLAY is non-nil, field values are returned in a way that is
suitable for display: in .bib files, @String abbreviations are
expanded, in .json files, values that are not strings are
converted to strings.  Furthermore, sequences of white space
characters (including newlines) are reduced to a single space.

...

```

 But I think I had misunderstood something about what the fields argument to `parsebib-parse` works, and @aikrahguzar's suggestion works like a charm. (It's a bit odd that passing arguments to the fields argument overrides essentially makes the function ignore the value of the display argument, but maybe I'm missing something. And in any case, that's an issue with `parsebib`.)

Thanks to you both for your help! I'm closing this here, but would you be interested in adding something to the Wiki or the README about making use of crossref? I could do this in the next few days if so.

**Comment by bdarcus** [comment]
We should then add that field to the default.

Care to do a PR?

**Comment by apc** [comment]
Happy to do so, but there's a small wrinkle as far as I can tell with making this the default. This is from `parsebib`'s  README: 

> Note that cross-references are resolved against the entries that appear in the buffer _above_ the current entry, and also against the entries in the hash table entries.

So, if you run `parsebib-parse` on a file containing this: 

```bib
@Book{key1,
  author       = {Smithers, Waylon},
  title        = {A Book},
  date         = 2020
}


@InBook{key2,
  title        = {A chapter},
  crossref     = {key1}
}
```

You will get the desired behavior. If instead you run it on a file with this as its content

```bib
@InBook{key2,
  title        = {A chapter},
  crossref     = {key1}
}


@Book{key1,
  author       = {Smithers, Waylon},
  title        = {A Book},
  date         = 2020
}

```

the result will not include author or year information for the item with `key1` as its key.

At the very least, then, we'd need to include some information about how to sort bib files in order for the crossref information to be used by `citar`. 

Unfortunately, there's an issue here, which is that in order for `bibtex.el` to make use of `crossref` information, you need to sort the file in the _opposite_ direction! That is, you need to set the value of `bibtex-maintained-sorted-entries` to `crossref`...

```
crossref     Sort entries alphabetically by keys unless an entry has a
                   crossref field.  These crossrefed entries are placed in
                   alphabetical order immediately preceding the main entry.
```

Any ideas how to get around this? The problem, granted, is only for those who use `bibtex.el` to edit their bib files and use `citar` to manage their references. I'm one of those, but I do not know how many others there are in that position.

**Comment by apc** [comment]
OK, I'm clearly misunderstanding something here in the documentation of `parsebib`. I just tried again with a file sorted by `bibtex.el` and another one sorted in the opposite direction and it seems to work fine. Will report back.

**Comment by apc** [comment]
OK, I need another pair of eyes. Could either one of you try running `parsebib-parse` on two files, one sorted so that the entry with the crossref field is first and one where it is last, and comparing the results? 

If you get the behavior I'm getting, where the order doesn't matter, then I'll make the PR. If I'm missing something and the order matters, then we'd want to make this behavior optional, or at least specify in the README that in order for it to work the bibliography files need to be sorted in a particular way...

**Comment by bdarcus** [comment]
I've been really busy with other things. It seems to me whether order matters or not should not impact the code here; that we should just include a link a prominent mention this is a restriction that comes from parsebib?

So the PR would simply be a one-line change in citar.el, and then a sentence or two in the README?

**Comment by apc** [comment]
Sure thing. I'm on it. I've tried a few times again and it doesn't look as if the order matters in the end. I don't know whether this is because `parsebib` is first generating and hash table with the entries when called with a nil `:entries` argument and then running over it again, as it were, but somehow the order doesn't seem to matter for our purposes. So I don't think there's need to add anything to the README. If you want me to anyway, I can add something like "note that use of crossref values may depend...". Just let me know in a comment on the PR.

**Comment by bdarcus** [comment]
Oops; forgot to link the commit and PR to this.

This issue is closed via #551.
